[WIP] Add VMDq support to intel_mp driver

takikawa opened the pull request:
This is a WIP pull request, so it's not intended to be merged in its current form. It's a first step in adding VMDq support to the `intel_mp` driver so that it can do VLAN and MAC address filtering.

Currently, this code tries to support VMDq with 32 pools with 4 RSS queues each (should it be configurable to 64 x 2?). A pool is selected with a `poolnum` argument (and the `poolnum` and `rxq` must match up according to how the 82599 assigns them).

Some of the code is taken from the current `intel` driver, especially for setting the MAC address. The code follows the initialization steps in 4.6.11.3.3 and 4.6.10.1.4 of the datasheet fairly closely in structure, but there are some missing things (like Tx configuration is not quite right probably).

Here are some TODO items and open questions that need to be resolved:

- [x] Support on i350 with no RSS (i210 not supported for VMDq)
  * decided to just error when VMDq is enabled on any card that isn't an 82599
- [ ] VLAN filtering and mirror settings implementation
- [ ] Testing everything much more
- [ ] Documentation update
- [ ] How should MTQC (multiple transmit queues) be configured in vmdq mode?
- [ ] 82599 supports only one redirection table (RETA register) for all VM pools. Should it configure RSS to always redirect to the min (or max?) of the Rx queues enabled per pool? Or always to 4?
- [ ] Only allow VMDq in follower instances when the main instance was run with VMDq on. (and vice versa, require VMDq if main instance used it)

takikawa committed:
First attempt at VMDq for intel_mp

takikawa committed:
Fix up VMDq mode so that it actually runs

takikawa committed:
First attempt at tests for VMDq mode

takikawa committed:
Try testing different MAC addrs with VMDq

takikawa committed:
Fix MAC address registration in VMDq mode

Avoid using index sets and just query the registers since it's
hard to share this state among the apps. Also do the MAC registration
in the right place.

takikawa committed:
Add comment explaining some RSS code

takikawa committed:
Enable RSS queues via PSRTYPE properly for VMDq

takikawa committed:
Use :bits to set the VMDq RSS mode properly

takikawa committed:
Remove RSS bit check for setting RETA

RSS is always enabled, so there shouldn't be a case in which
this is a no-op anyway. Also the check is complicated by the fact
that RSS can be enabled in many ways (with DCB, VMDq, etc.).

takikawa committed:
Fix RTTPCS bit set that had a wrong length arg

takikawa committed:
Adjust test to test that both vmdq pools get pkts

takikawa committed:
Move MAC pool enable code and make it cross-NIC

Also add code to unset MAC on stop()

takikawa committed:
Adjust the VMDq initialization assertions

takikawa committed:
Error when the max number of MAC addresses is reached

takikawa committed:
Add code for enabling VLAN filtering/tagging

This code is mostly copied from the intel10g.lua driver.
Also add a new VLAN test and adjust old tests as needed.

takikawa committed:
Refactor vmdq test script

The previous iteration was unreliable for unknown reasons. This
refactoring should be clearer and run more reliably.

takikawa committed:
Enable VLAN takedown code for VMDq

teknico commented on src/apps/intel_mp/intel_mp.lua:
```diff
@@ -953,9 +1150,91 @@ function Intel82599:init ()
+      self.r.PFQDE(bor(lshift(1,16), lshift(i,8)))
+   end
+
+   -- clear RTTDT1C, PFVLVF for all pools, set them later
```
No PFVLVF in the code below.

takikawa replied:
It's set in `add_receive_VLAN` I think (which is mostly copied from `intel10g` with some changes).

teknico commented on src/apps/intel_mp/intel_mp.lua:
```diff
@@ -375,6 +429,15 @@ function Intel:init_rx_q ()
+      -- packet splitting none, enable 4 RSS queues per pool
+      self.r.PSRTYPE[self.poolnum](bits { RQPL=30 })
+      -- multicast promiscuous, broadcast accept, accept untagged pkts
+      self.r.PFVML2FLT[self.poolnum]:set(bits { MPE=28, BAM=27, AUPE=24 })
```
Do untagged incoming packets get copied to all pools? Are all pools able to send untagged packets?

takikawa replied:
I believe this means untagged packets get sent to whichever app instance matches the MAC on the packet (I copied this behavior from `intel10g` I think). For sending I think it depends on if we set VLAN tag insertion on.

teknico commented on src/apps/intel_mp/intel_mp.lua:
```diff
@@ -953,9 +1150,91 @@ function Intel82599:init ()
+   -- QDE bit = 0 for all queues
+   for i = 0, 127 do
+      self.r.PFQDE(bor(lshift(1,16), lshift(i,8)))
+   end
```
Setting PFQDE is not mentioned in section 4.6.11.3.3 "DCB-Off, VT-On".

takikawa replied:
It's mentioned in 4.6.11.3.1:

> Queue Drop Enable (PFQDE) — In SR-IO the QDE bit should be set to 1b in the
PFQDE register for all queues. In VMDq mode, the QDE bit should be set to 0b for
all queues.

I think 4.6.11.3.1 is relevant since 4.6.11.3.3 says:

> Set the configuration bits as specified in Section 4.6.11.3.1 with the following
exceptions:

teknico commented on src/apps/intel_mp/intel_mp.lua:
```diff
@@ -953,9 +1150,91 @@ function Intel82599:init ()
+   self.r.RTTDCS:set(bits { VMPAC=1, BDPM=22 })
+   self.r.RTTPCS:clr(bits { TPPAC=5, TPRM=8 })
+   -- set RTTPCS.ARBD
+   self.r.RTTPCS:bits(22, 10, 0x244)
```
Why are bits 22 and 10 set? I don't see them in section 8.2.3.10.3 .

takikawa replied:
This syntax is a bit confusing perhaps, but this sets bits 22 to 32 to `0x244` I think (i.e., the `ARBD` field). Which is part of the setup (see page 185).

teknico commented on src/apps/intel_mp/intel_mp.lua:
```diff
@@ -953,9 +1150,91 @@ function Intel82599:init ()
+   end
+
+   -- disable TC arbitrations, enable packet buffer free space monitor
+   self.r.RTTDCS:clr(bits { TDPAC=0, TDRM=4, BPBFSM=23 })
```
Field BPBFSM of section 8.2.3.10.2 says: "This bit must be cleared in DCB mode or SR-IOV mode." But we're using VMDq mode, not SR-IOV, right?

takikawa replied:
Yes, I'm not sure whether it needs to be cleared in VMDq mode. I cleared this based on page 184, where it says to set the bits to `TDPAC=0b, VMPAC=1b, TDRM=0b, BDPM=1b, BPBFSM=0b` for DCB-off, VT-on.

teknico reviewed (commented):
> Support on i350 with no RSS (i210 not supported for VMDq)

Why no RSS? The i350 does support RSS.

Since we're enabling VF Receive Enable (PFVFRE register), maybe we should also enable VF Transmit
Enable (PFVFTE register), section 4.2.2.2 .

Is the RQPL field of section 8.2.3.7.4 incorrect where it says: "Valid values are zero, 0001b and 0010b."? Since bits are 31:29, they should say zero, 001b and 010b.

wingo commented:
Is there any reason for this PR to be against `lwaftr` ?  Maybe that's just part of the WIP status or something, but eventually it should be against `SnabbCo:master`.  If you want to develop the WIP PR against `master` that would be fine but I also understand if you feel like that would bother people.  Only thing is, drive-by reviews will come more easily against `SnabbCo:master` and they can be valuable.

takikawa commented:
@teknico My understanding is that the i350 supports VMDq and RSS separately, but not together. I agree about the transmit (focused so far on getting receive to work, transmit is possibly non-working). Re: `RQPL`, I think the code is setting the relevant bit in `init_rx_q` (where it says `RQPL=30`, so `0010b` I think) but maybe I'm overlooking something?

@wingo The only reason is I thought maybe we should get it in a somewhat usable state before PRing it to the main snabb repo, but I'm also open to just putting it there if you all think that's for the best.

wingo commented:
@takikawa no change needed, just wanted to make sure we were on the same page regarding final destination of the pr as well as the drive-by review factor.  happy hacking :)

takikawa committed:
Refactor intel_mp recv test scripts to reduce duplication

takikawa committed:
Enable VLAN tag stripping in VMDq mode

Also add new test for VLAN stripping

takikawa committed:
Adjust VLAN test to also test tag insertion

takikawa committed:
VMDq settings for transmit queues

takikawa committed:
Add a test for VLAN tag insertion & VMDq tx

takikawa committed:
Set registers needed to make Tx test work

 * PFVFTE to enable Tx from appropriate VFs
 * RTTD1TC for bandwidth allocation algorithm for Tx

takikawa committed:
Fix assertion for Rx queue num to allow nil

takikawa committed:
Adjust intel_mp to error with VMDq on non-82599

takikawa committed:
Remove unnecessary helper function

takikawa committed:
Fix permissions on two intel_mp tests

takikawa committed:
Fix 1q vmdq test and testrecv.lua

The former wasn't updated for an API change and the latter
was missing a module declaration.

takikawa committed:
Adjust VLAN test to avoid testing tag insertion

The tag insertion part of this test didn't work because the
MAC didn't match when the app tried to resend the packet. It's
not necessary since the transmit test covers this anyway.

takikawa committed:
Add waits to avoid race conditions in vmdq tests

takikawa committed:
Add documentation for new vmdq-related parameters

takikawa committed:
Refactor vmdq-related assertions

Commit:
Make intel_mp tests run even when either 1G or 10G cards are not present

takikawa committed:
Merge pull request #2 from teknico/vmdq-10g-only-tests

WIP: make intel_mp tests run even when either 1G or 10G cards are not present

takikawa committed:
Adjust VMDq test for 82599 to be more robust

There are multiple ids for 82599 variants that may
be added later

takikawa committed:
Add VMDq consistency check for intel_mp apps

Makes sure that all apps on a given NIC either use or don't
use VMDq uniformly

takikawa commented:
These changes are now being PRed to the upstream at https://github.com/snabbco/snabb/pull/1157 so I'm going to close this PR for now. For future work on `intel_mp`, I'll push to another branch here and open a new PR so that the CI runs.
